storage/merkle/proof.go

```diff
@@ -0,0 +1,133 @@
+// the constructed root hash with the given one.
+// if the proof is valid it returns true and false otherwise
+func (p *Proof) Verify(expectedRootHash []byte) (bool, error) {
+
+	// first check the key size to be smaller than max allowed
+	if len(p.Key) > maxKeyLength {
+		return false, NewErrorMalformedProoff("key length is larger than max key lenght allowed (%d > %d)", len(p.Key), maxKeyLength)
+	}
+
+	// number of steps
+	steps := len(p.SkipBits) + len(p.InterimHashes)
+	// number of steps should be smaller than max key length
+	if steps > maxKeyLength {
+		return false, NewErrorMalformedProoff("length of SkipBits pluse length of InterimHashes is larger than max key lenght allowed (%d > %d)", steps, maxKeyLength)
+	}
+	// number of steps should be smaller than number of bits in the IsAShortNode
+	if steps > len(p.IsAShortNode)*8 {
+		return false, NewErrorMalformedProoff("the length of IsAShortNode doesn't match the length of SkipBits and InterimHashes")
+	}
+
+	// an index to consume interim hashes from the last element to the first element
+	interimHashIndex := len(p.InterimHashes) - 1
+	// an index to consume shortBits from the last element to the first element
+	skipBitIndex := len(p.SkipBits) - 1
+
+	// keyIndex keeps track of the largest index of the key that is unchecked.
+	// note that traverse bottom up here, so we start with the largest key index
+	// build hashes until we reach to the root.
+	keyIndex := len(p.InterimHashes)
+	for _, sc := range p.SkipBits {
+		keyIndex += int(sc)
+	}
+	keyIndex-- // consider index starts from zero
+
+	if len(p.Key)*8 != keyIndex+1 {
+		return false, NewErrorMalformedProoff("key length doesn't match the length of SkipBits and InterimHashes")
+	}
+
+	// compute the hash value of the leaf
+	currentHash := computeLeafHash(p.Value)
+
+	// for each step (level from bottom to top) check if its a full node or a short node and compute the
+	// hash value accordingly; for full node having the sibling hash helps to compute the hash value
+	// of the next level, for short nodes compute the hash using the common path constructed based on
+	// the given short count
+	for isAShortNodeIndex := steps - 1; isAShortNodeIndex >= 0; isAShortNodeIndex-- {
+
+		// Full node
+		if bitutils.ReadBit(p.IsAShortNode, isAShortNodeIndex) == 0 {
+
+			// read and pop the sibling hash value from InterimHashes
+			if interimHashIndex < 0 {
+				return false, NewErrorMalformedProoff("no more InterimHashes available to read")
+			}
+			sibling := p.InterimHashes[interimHashIndex]
+			interimHashIndex--
+
+			// based on the bit at pathIndex of the key compute the hash
+			if bitutils.ReadBit(p.Key, keyIndex) == 0 { // left branching
+				currentHash = computeFullHash(currentHash, sibling)
+			} else {
+				currentHash = computeFullHash(sibling, currentHash) // right branching
+			}
+
+			// decrement the keyindex by 1
+			keyIndex--
+
+			continue
```

Comment:
You are describing _what_ is happening,  i.e. `keyIndex` is decremented, which is trivially apparent from the code. But you are not describing why (which might deserve a comment)

```suggestion
			// move to the parent vertex along the path
			keyIndex--
			continue
```

Reply:
fixed